### PR TITLE
fix(modem): Fixed mode transitions between any state and UNDEF

### PR DESCRIPTION
**Pull request by david-cermak**
Closes https://github.com/espressif/esp-protocols/issues/320

### CODE REVIEW

**Review by gabsuren — APPROVED** [review]

File: components/esp_modem/src/esp_modem_dte.cpp [hunk 1]
```diff
@@ -151,10 +151,14 @@ command_result DTE::command(const std::string &cmd, got_line_cb got_line, uint32
...
 bool DTE::exit_cmux()
 {
+    if (!cmux_term) {
+        return false;
+    }
     if (!cmux_term->deinit()) {
         return false;
     }
     exit_cmux_internal();
+    cmux_term = nullptr;
```
Comment (euripedesrocha) [inline components/esp_modem/src/esp_modem_dte.cpp:161]:
```suggestion
    cmux_term.reset();
```

Comment (euripedesrocha) [reply components/esp_modem/src/esp_modem_dte.cpp:161]: Using reset makes the intention clear.

File: components/esp_modem/src/esp_modem_dte.cpp [hunk 2]
```diff
@@ -174,6 +178,9 @@ void DTE::exit_cmux_internal()
 
 bool DTE::setup_cmux()
 {
+    if (cmux_term) {
```
Comment (euripedesrocha) [inline components/esp_modem/src/esp_modem_dte.cpp:181]: Shall we provide some debug logs here?

**Review by euripedesrocha — APPROVED** [review]
LGTM. Left a few minor comments.

File: components/esp_modem/src/esp_modem_dte.cpp (hunk 1 above)
Comment (david-cermak) [reply components/esp_modem/src/esp_modem_dte.cpp]: Thanks! changed

File: components/esp_modem/src/esp_modem_dte.cpp (hunk 2 above)
Comment (david-cermak) [reply components/esp_modem/src/esp_modem_dte.cpp:181]: Makes sense, added `ESP_LOGE(...)`

**Comment by fdkz** [comment]
Doesn't seem to fully work as intended with SIM7600.

Test case:

dce->set_mode(esp_modem::modem_mode::CMUX_MODE); // OK
physically reset modem
dce->set_mode(esp_modem::modem_mode::UNDEF); // OK
dce->set_mode(esp_modem::modem_mode::CMUX_MODE); // FAIL

esp_modem starts to send +++ and AT wrapped:
`\xF9\t\xEF\x07+++5\xF9`
`\xF9\t\xEF\x07AT\r5\xF9`

and doesn't recognize the unwrapped response by the modem:
`AT\r\r\nOK\r\n`

**Comment by david-cermak** [comment]
> Doesn't seem to fully work as intended with SIM7600.
> 
> Test case:
> 
> dce->set_mode(esp_modem::modem_mode::CMUX_MODE); // OK physically reset modem dce->set_mode(esp_modem::modem_mode::UNDEF); // OK dce->set_mode(esp_modem::modem_mode::CMUX_MODE); // FAIL
> 
> esp_modem starts to send +++ and AT wrapped: `\xF9\t\xEF\x07+++5\xF9` `\xF9\t\xEF\x07AT\r5\xF9`
> 
> and doesn't recognize the unwrapped response by the modem: `AT\r\r\nOK\r\n`

That's not how the `UNDEF` mode is supposed to work. It just a meta-state indicating that all transitions are allowed, not that they'd necessarily succeed (nor it is a synonym to INIT or RESET state).

What you describe in your test case is more like a situation  suitable for mode detection (another planned feature), in other words: We don't know what state the modem is in, so we check how it responds to all protocols and set the mode accordingly.

**Pull request merged**